src/aggregator/client/writer.go

```diff
@@ -93,39 +93,43 @@ func newInstanceWriter(instance placement.Instance, opts Options) instanceWriter
 	return w
 }
 
-func (w *writer) Write(shard uint32, payload payloadUnion) error {
+func (w *writer) Write(shard uint32, payload payloadUnion) (int, error) {
+	// Read lock to check if the writer is closed and try to get the encoder.
 	w.RLock()
 	if w.closed {
 		w.RUnlock()
-		return errInstanceWriterClosed
+		return 0, errInstanceWriterClosed
 	}
+
 	encoder, exists := w.encodersByShard[shard]
+	w.RUnlock()
+
 	if exists {
-		err := w.encodeWithLock(encoder, payload)
-		w.RUnlock()
-		return err
+		// If the encoder exists, encode without acquiring the write lock.
+		return w.encodeWithLock(encoder, payload)
 	}
-	w.RUnlock()
 
+	// Acquire the write lock if the encoder doesn't exist.
 	w.Lock()
+	defer w.Unlock()
```

Comment:
`defer()` prevents inlining in all kinds of weird ways. can you avoid relying on that here.